File: app/src/main/java/net/emilla/lang/phrase/Dice.java
package net.emilla.lang.phrase;

import androidx.annotation.Nullable;

import net.emilla.util.Hashes;

import java.util.random.RandomGenerator;

public final class Dice implements Comparable<Dice> {

    private int mCount;
    private final int mFaces;

    public Dice(int count, int faces) {
        mCount = count;
        mFaces = faces;
    }

    public void add(int count) {
        mCount += count;
    }

    public int count() {
        return mCount;
    }

    public int roll(RandomGenerator rand) {
        if (mFaces == 1) {
            return mCount;
        }

        int result = 0;
        if (mCount >= 0) {
            for (int i = 0; i < mCount; ++i) {
                result += rollOne(rand);
            }
        } else {
            for (int i = 0; i < -mCount; ++i) {
                result -= rollOne(rand);
            }
        }

        return result;
    }

    private int rollOne(RandomGenerator rand) {
        return rand.nextInt(mFaces) + 1;
    }

    @Override
    public int compareTo(Dice other) {
        return Integer.compare(mFaces, other.mFaces);
    }

    @Override
    public boolean equals(@Nullable Object other) {
        return other instanceof Dice dice
            && mFaces == dice.mFaces;
    }

    @Override
    public int hashCode() {
        return Hashes.one(mFaces);
    }

}

File: app/src/main/java/net/emilla/lang/phrase/Dice.kt
package net.emilla.lang.phrase

import net.emilla.util.hash1
import java.util.Random

class Dice(count: Int, val faces: Int) : Comparable<Dice> {
    var count = count @JvmName("count") get
        private set

    fun add(count: Int) {
        this.count += count
    }

    fun roll(rand: Random): Int {
        if (faces == 1) return count

        var result = 0
        if (count >= 0) repeat(count) {
            result += rand.nextInt(faces) + 1
        } else repeat(count) {
            result -= rand.nextInt(faces) + 1
        }

        return result
    }

    override fun compareTo(other: Dice) = faces - other.faces
    override fun equals(other: Any?) = this === other || other is Dice && faces == other.faces
    override fun hashCode() = hash1(faces)
}
